kikuchipy/projections/lambert_projection.py

```diff
@@ -59,11 +59,15 @@ def project(cls, v: Union[Vector3d, np.ndarray]) -> np.ndarray:
         lambert[..., 0] = np.where(
             abs_yx,
             sign_x * sqrt_z * sqrt_pi_half,
-            sign_y * sqrt_z * (two_over_sqrt_pi * np.arctan(x / y)),
+            sign_y
+            * sqrt_z
+            * (two_over_sqrt_pi * np.arctan(np.divide(x, y, where=y != 0))),
```

Comment:
If I'm reading the NumPy documentation (https://numpy.org/doc/stable/reference/generated/numpy.divide.html) correctly, the output array from numpy.divide will have uninitialized values where the condition is false, right? Are the uninitialized values always zero?

Just want to be sure of this before merging.

Reply:
There will be uninitialized values that can differ from zero.

Let's consider the edge cases:

[x, y, z] | y <= x | x-calc | y-calc|
---|---|---|---|
[1, 0, 0] | True | Normal | np.divide -> y/x = 0/1 |
[0, 1, 0] | False | np.divide -> x/y = 0/1 | Normal |
[0, 0, 1] | True | Normal | np.divide -> y/x = 0/0 |

So it should only give uninitialized values for [0, 0, 1], as in the other cases it doesn't return any uninitialized values. 

When the input is [0, 0, 1]:

```python
>>> sqrt_z = np.sqrt(2 * (1 - abs(z)))
>>> sqrt_z == 0
True
``` 
on line 46 should be equal to 0, and thus the y calculation (line 66) where it gives an uninitialized value 

```python
              sign_x
            * sqrt_z
            * (two_over_sqrt_pi * np.arctan(np.divide(y, x, where=x != 0)))
```
should also give 0. 

Running the following test program:
```python
import numpy as np
from kikuchipy.projections.lambert_projection import LambertProjection

xyz = np.array(([0, 0, 1], [0, 0, 1], [0, 0, 1]))

x = xyz[..., 0]
y = xyz[..., 1]
z = xyz[..., 2]

div_xy = np.divide(x, y, where=y != 0)
div_yx = np.divide(y, x, where=x != 0)
print("X - values")
print(x)
print("Y - values")
print(y)
print("X / Y")
print(div_xy)
print("Y / X")
print(div_yx)

print("Inside Lambert")
xy = LambertProjection.project(xyz)
print("---")
print(xy[..., 0])
print(xy[..., 1])
print("---")
```


```
X - values
[0 0 0]
Y - values
[0 0 0]
X / Y
[1. 1. 1.]
Y / X
[2.47379808e-091 1.03983890e-152 9.08367206e+223]
Inside Lambert
---
[0. 0. 0.]
[0. 0. 0.]
---
```


So to sum it up:
Yes there will be uninitialized values, however I do not believe they cause any problems as they get multiplied by zero whenever they are returned... maybe this is undefined behavior?

Reply:
Thanks a lot for checking this!

According to EMsoft (https://github.com/EMsoft-org/EMsoft/blob/develop/Source/EMsoftLib/Lambert.f90#L365), which makes sense to me, `LambertProjection.project()` should return (X, Y) = (0, 0) in the cases where:
* | 1 - (x^2 + y^2 + z^2) | > eps, where eps is set to `np.finfo(np.float).eps`, which is 
np.finfo(np.float).eps 220446049250313e-16 on my machine
* | z | = 1

The first one is taken care of by us, the second is not, which results in NaNs, as you say. I propose these changes:
```python
    @classmethod
    def project(cls, v: Union[Vector3d, np.ndarray]) -> np.ndarray:
        """Convert (n, 3) vector from Cartesian to the Lambert projection."""
        if isinstance(v, Vector3d):
            w = v.unit.data
            x = w[..., 0]
            y = w[..., 1]
            z = w[..., 2]
        else:
            w = np.atleast_2d(v)
            norm = np.linalg.norm(w, axis=-1)[..., np.newaxis]
            w = w / norm
            x = w[..., 0]
            y = w[..., 1]
            z = w[..., 2]

        # Arrays used in both setting X and Y
        sqrt_z = np.sqrt(2 * (1 - abs(z)))
        sign_x = np.sign(x)
        sign_y = np.sign(y)
        abs_yx = abs(y) <= abs(x)

        # Reusable constants
        sqrt_pi = np.sqrt(np.pi)
        sqrt_pi_half = sqrt_pi / 2
        two_over_sqrt_pi = 2 / sqrt_pi

        # Ensure (0, 0) is returned where |z| = 1
        lambert = np.zeros(x.shape + (2,), dtype=x.dtype)
        z_not_one = abs(z) != 1

        # Equations (10a) and (10b) from Callahan and De Graef (2013)
        lambert[z_not_one, 0] = np.where(
            abs_yx,
            sign_x * sqrt_z * sqrt_pi_half,
            sign_y
            * sqrt_z
            * (two_over_sqrt_pi * np.arctan(x / y)),
        )[z_not_one]
        lambert[z_not_one, 1] = np.where(
            abs_yx,
            sign_x
            * sqrt_z
            * (two_over_sqrt_pi * np.arctan(y / x)),
            sign_y * sqrt_z * sqrt_pi_half,
        )[z_not_one]

        return lambert
```

This results in
```python
>>> import numpy as np
>>> from kikuchipy.projections.lambert_projection import LambertProjection
>>> xyz = np.array([[0, 0, 1], [0, 1, 0], [2, 0, 0], [0, 0, -3]])
>>> print(LambertProjection.project(xyz))
[[0.         0.        ]
 [0.         1.25331414]
 [1.25331414 0.        ]
 [0.         0.        ]]
```

Reply:
I think it is safe to ignore `RuntimeWarning` in true_divide:

```python
      with np.errstate(divide="ignore", invalid="ignore"):
            lambert[z_not_one, 0] = np.where(
                abs_yx,
                sign_x  sqrt_z  sqrt_pi_half,
                sign_y
                * sqrt_z
                 (two_over_sqrt_pi  np.arctan(x / y)),
            )[z_not_one]
            lambert[z_not_one, 1] = np.where(
                abs_yx,
                sign_x
                * sqrt_z
                 (two_over_sqrt_pi  np.arctan(y / x)),
                sign_y  sqrt_z  sqrt_pi_half,
            )[z_not_one]
```